### PR TITLE
Fix anon const def-creation when macros are involved

### LINKED ISSUE #128016: ICE: `adding a def'n for node-id NodeId(18) and def kind AnonConst but a previous def'n exists`
auto-reduced (treereduce-rust):
````rust
macro_rules! len {
    () => {
        target
    };
}

fn main() {
    let val: [str; len!()] = [];
}


````

original:
````rust
#![deny("{}", equal(&1, &2))]

macro_rules! len {
    () => (target) //~  ERROR trailing semicolon
                  //~| WARN this was previously accepted
}

fn main() {
    let val: [str; len!()] = [];
}

````

Version information
````
rustc 1.81.0-nightly (2e6fc4254 2024-07-20)
binary: rustc
commit-hash: 2e6fc425411280c658d852813de01c5157255088
commit-date: 2024-07-20
host: x86_64-unknown-linux-gnu
release: 1.81.0-nightly
LLVM version: 18.1.7

````

Command:
`/home/matthias/.rustup/toolchains/master/bin/rustc `

<details><summary><strong>Program output</strong></summary>
<p>

```
error[E0425]: cannot find value `target` in this scope
 --> /tmp/icemaker_global_tempdir.1khAqktZak0k/rustc_testrunner_tmpdir_reporting.F2c9o1Y4kXrG/mvce.rs:3:9
  |
3 |         target
  |         ^^^^^^ not found in this scope
...
8 |     let val: [str; len!()] = [];
  |                    ------ in this macro invocation
  |
  = note: this error originates in the macro `len` (in Nightly builds, run with -Z macro-backtrace for more info)

thread 'rustc' panicked at compiler/rustc_ast_lowering/src/lib.rs:503:9:
adding a def'n for node-id NodeId(18) and def kind AnonConst but a previous def'n exists: DefKey { parent: Some(DefIndex(4)), disambiguated_data: DisambiguatedDefPathData { data: AnonConst, disambiguator: 0 } }
stack backtrace:
   0:     0x7e8412c083a5 - std::backtrace_rs::backtrace::libunwind::trace::h160e615c518ecb01
                               at /rustc/2e6fc425411280c658d852813de01c5157255088/library/std/src/../../backtrace/src/backtrace/libunwind.rs:116:5
   1:     0x7e8412c083a5 - std::backtrace_rs::backtrace::trace_unsynchronized::he30c6d2f98d2efdc
                               at /rustc/2e6fc425411280c658d852813de01c5157255088/library/std/src/../../backtrace/src/backtrace/mod.rs:66:5
   2:     0x7e8412c083a5 - std::sys::backtrace::_print_fmt::hb87cad28f2e90a0b
                               at /rustc/2e6fc425411280c658d852813de01c5157255088/library/std/src/sys/backtrace.rs:65:5
   3:     0x7e8412c083a5 - <std::sys::backtrace::BacktraceLock::print::DisplayBacktrace as core::fmt::Display>::fmt::h362429114acc76fb
                               at /rustc/2e6fc425411280c658d852813de01c5157255088/library/std/src/sys/backtrace.rs:40:26
   4:     0x7e8412c57eab - core::fmt::rt::Argument::fmt::h2dd3e5813b018ccc
                               at /rustc/2e6fc425411280c658d852813de01c5157255088/library/core/src/fmt/rt.rs:173:76
   5:     0x7e8412c57eab - core::fmt::write::haceebbcd70688c9b
                               at /rustc/2e6fc425411280c658d852813de01c5157255088/library/core/src/fmt/mod.rs:1182:21
   6:     0x7e8412bfce7f - std::io::Write::write_fmt::h7c7b83fdf5105f1d
                               at /rustc/2e6fc425411280c658d852813de01c5157255088/library/std/src/io/mod.rs:1827:15
   7:     0x7e8412c0ab91 - std::sys::backtrace::BacktraceLock::print::hd97590b35fdf3d35
                               at /rustc/2e6fc425411280c658d852813de01c5157255088/library/std/src/sys/backtrace.rs:43:9
   8:     0x7e8412c0ab91 - std::panicking::default_hook::{{closure}}::h587635c60073753d
                               at /rustc/2e6fc425411280c658d852813de01c5157255088/library/std/src/panicking.rs:269:22
   9:     0x7e8412c0a86c - std::panicking::default_hook::h41bedcdd5e6a0e44
                               at /rustc/2e6fc425411280c658d852813de01c5157255088/library/std/src/panicking.rs:296:9
  10:     0x7e840f024e1a - std[f9e60369e5bdbf77]::panicking::update_hook::<alloc[f72c5bad8b9f51f]::boxed::Box<rustc_driver_impl[23129d3f9f762646]::install_ice_hook::{closure#0}>>::{closure#0}
  11:     0x7e8412c0b55f - <alloc::boxed::Box<F,A> as core::ops::function::Fn<Args>>::call::h7651476b4c558d70
                               at /rustc/2e6fc425411280c658d852813de01c5157255088/library/alloc/src/boxed.rs:2084:9
  12:     0x7e8412c0b55f - std::panicking::rust_panic_with_hook::heee9f152ffb5d0c8
                               at /rustc/2e6fc425411280c658d852813de01c5157255088/library/std/src/panicking.rs:808:13
  13:     0x7e8412c0b187 - std::panicking::begin_panic_handler::{{closure}}::h808b135ae5150977
                               at /rustc/2e6fc425411280c658d852813de01c5157255088/library/std/src/panicking.rs:674:13
  14:     0x7e8412c08869 - std::sys::backtrace::__rust_end_short_backtrace::h3c8a02e9918c1a26
                               at /rustc/2e6fc425411280c658d852813de01c5157255088/library/std/src/sys/backtrace.rs:168:18
  15:     0x7e8412c0ae14 - rust_begin_unwind
                               at /rustc/2e6fc425411280c658d852813de01c5157255088/library/std/src/panicking.rs:665:5
  16:     0x7e8412c54463 - core::panicking::panic_fmt::h124ff2929e2b97c8
                               at /rustc/2e6fc425411280c658d852813de01c5157255088/library/core/src/panicking.rs:74:14
  17:     0x7e8412045276 - <rustc_ast_lowering[d5c02056e301041c]::LoweringContext>::create_def.cold
  18:     0x7e8410ccdfdd - <rustc_ast_lowering[d5c02056e301041c]::LoweringContext>::lower_anon_const_to_anon_const
  19:     0x7e8410ccdc72 - <rustc_ast_lowering[d5c02056e301041c]::LoweringContext>::lower_anon_const_to_const_arg_direct
  20:     0x7e8410ccd9f3 - <rustc_ast_lowering[d5c02056e301041c]::LoweringContext>::lower_array_length
  21:     0x7e8410cf9b8f - <rustc_ast_lowering[d5c02056e301041c]::LoweringContext>::lower_ty_direct
  22:     0x7e8410cae238 - <rustc_ast_lowering[d5c02056e301041c]::LoweringContext>::lower_block_noalloc
  23:     0x7e8410cd4d10 - <rustc_ast_lowering[d5c02056e301041c]::LoweringContext>::lower_maybe_coroutine_body
  24:     0x7e8410cee69f - <rustc_ast_lowering[d5c02056e301041c]::LoweringContext>::lower_item_kind
  25:     0x7e8410cd215f - <rustc_ast_lowering[d5c02056e301041c]::item::ItemLowerer>::lower_node
  26:     0x7e8410ccfb79 - rustc_ast_lowering[d5c02056e301041c]::lower_to_hir
  27:     0x7e84114c5260 - rustc_query_impl[7076d3a069d488ea]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[7076d3a069d488ea]::query_impl::hir_crate::dynamic_query::{closure#2}::{closure#0}, rustc_middle[f5371164b7025427]::query::erase::Erased<[u8; 8usize]>>
  28:     0x7e84114c2302 - rustc_query_system[eb2b374bcd8a9022]::query::plumbing::try_execute_query::<rustc_query_impl[7076d3a069d488ea]::DynamicConfig<rustc_query_system[eb2b374bcd8a9022]::query::caches::SingleCache<rustc_middle[f5371164b7025427]::query::erase::Erased<[u8; 8usize]>>, false, false, false>, rustc_query_impl[7076d3a069d488ea]::plumbing::QueryCtxt, false>
  29:     0x7e84114c1f1e - rustc_query_impl[7076d3a069d488ea]::query_impl::hir_crate::get_query_non_incr::__rust_end_short_backtrace
  30:     0x7e8410e9a064 - rustc_query_impl[7076d3a069d488ea]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[7076d3a069d488ea]::query_impl::hir_attrs::dynamic_query::{closure#2}::{closure#0}, rustc_middle[f5371164b7025427]::query::erase::Erased<[u8; 8usize]>>
  31:     0x7e8410e9961f - rustc_query_system[eb2b374bcd8a9022]::query::plumbing::try_execute_query::<rustc_query_impl[7076d3a069d488ea]::DynamicConfig<rustc_query_system[eb2b374bcd8a9022]::query::caches::VecCache<rustc_hir[2699c16870db6088]::hir_id::OwnerId, rustc_middle[f5371164b7025427]::query::erase::Erased<[u8; 8usize]>>, false, false, false>, rustc_query_impl[7076d3a069d488ea]::plumbing::QueryCtxt, false>
  32:     0x7e8410e9938f - rustc_query_impl[7076d3a069d488ea]::query_impl::hir_attrs::get_query_non_incr::__rust_end_short_backtrace
  33:     0x7e8410878006 - <rustc_middle[f5371164b7025427]::hir::map::Map>::attrs
  34:     0x7e84117af401 - rustc_passes[141401c4575c1f66]::entry::entry_fn
  35:     0x7e84117af398 - rustc_query_impl[7076d3a069d488ea]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[7076d3a069d488ea]::query_impl::entry_fn::dynamic_query::{closure#2}::{closure#0}, rustc_middle[f5371164b7025427]::query::erase::Erased<[u8; 12usize]>>
  36:     0x7e84117af36f - <rustc_query_impl[7076d3a069d488ea]::query_impl::entry_fn::dynamic_query::{closure#2} as core[c927cdc4da4eb0de]::ops::function::FnOnce<(rustc_middle[f5371164b7025427]::ty::context::TyCtxt, ())>>::call_once
  37:     0x7e84117aec88 - rustc_query_system[eb2b374bcd8a9022]::query::plumbing::try_execute_query::<rustc_query_impl[7076d3a069d488ea]::DynamicConfig<rustc_query_system[eb2b374bcd8a9022]::query::caches::SingleCache<rustc_middle[f5371164b7025427]::query::erase::Erased<[u8; 12usize]>>, false, false, false>, rustc_query_impl[7076d3a069d488ea]::plumbing::QueryCtxt, false>
  38:     0x7e84117aea29 - rustc_query_impl[7076d3a069d488ea]::query_impl::entry_fn::get_query_non_incr::__rust_end_short_backtrace
  39:     0x7e8410b0ffb2 - rustc_interface[635a4d36e766d1e0]::passes::analysis
  40:     0x7e8410b0fdc7 - rustc_query_impl[7076d3a069d488ea]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[7076d3a069d488ea]::query_impl::analysis::dynamic_query::{closure#2}::{closure#0}, rustc_middle[f5371164b7025427]::query::erase::Erased<[u8; 1usize]>>
  41:     0x7e84114c3725 - rustc_query_system[eb2b374bcd8a9022]::query::plumbing::try_execute_query::<rustc_query_impl[7076d3a069d488ea]::DynamicConfig<rustc_query_system[eb2b374bcd8a9022]::query::caches::SingleCache<rustc_middle[f5371164b7025427]::query::erase::Erased<[u8; 1usize]>>, false, false, false>, rustc_query_impl[7076d3a069d488ea]::plumbing::QueryCtxt, false>
  42:     0x7e84114c348f - rustc_query_impl[7076d3a069d488ea]::query_impl::analysis::get_query_non_incr::__rust_end_short_backtrace
  43:     0x7e84114002c5 - rustc_interface[635a4d36e766d1e0]::interface::run_compiler::<core[c927cdc4da4eb0de]::result::Result<(), rustc_span[8f3a5487dfe29266]::ErrorGuaranteed>, rustc_driver_impl[23129d3f9f762646]::run_compiler::{closure#0}>::{closure#1}
  44:     0x7e8411394189 - std[f9e60369e5bdbf77]::sys::backtrace::__rust_begin_short_backtrace::<rustc_interface[635a4d36e766d1e0]::util::run_in_thread_with_globals<rustc_interface[635a4d36e766d1e0]::util::run_in_thread_pool_with_globals<rustc_interface[635a4d36e766d1e0]::interface::run_compiler<core[c927cdc4da4eb0de]::result::Result<(), rustc_span[8f3a5487dfe29266]::ErrorGuaranteed>, rustc_driver_impl[23129d3f9f762646]::run_compiler::{closure#0}>::{closure#1}, core[c927cdc4da4eb0de]::result::Result<(), rustc_span[8f3a5487dfe29266]::ErrorGuaranteed>>::{closure#0}, core[c927cdc4da4eb0de]::result::Result<(), rustc_span[8f3a5487dfe29266]::ErrorGuaranteed>>::{closure#0}::{closure#0}, core[c927cdc4da4eb0de]::result::Result<(), rustc_span[8f3a5487dfe29266]::ErrorGuaranteed>>
  45:     0x7e8411393f3a - <<std[f9e60369e5bdbf77]::thread::Builder>::spawn_unchecked_<rustc_interface[635a4d36e766d1e0]::util::run_in_thread_with_globals<rustc_interface[635a4d36e766d1e0]::util::run_in_thread_pool_with_globals<rustc_interface[635a4d36e766d1e0]::interface::run_compiler<core[c927cdc4da4eb0de]::result::Result<(), rustc_span[8f3a5487dfe29266]::ErrorGuaranteed>, rustc_driver_impl[23129d3f9f762646]::run_compiler::{closure#0}>::{closure#1}, core[c927cdc4da4eb0de]::result::Result<(), rustc_span[8f3a5487dfe29266]::ErrorGuaranteed>>::{closure#0}, core[c927cdc4da4eb0de]::result::Result<(), rustc_span[8f3a5487dfe29266]::ErrorGuaranteed>>::{closure#0}::{closure#0}, core[c927cdc4da4eb0de]::result::Result<(), rustc_span[8f3a5487dfe29266]::ErrorGuaranteed>>::{closure#1} as core[c927cdc4da4eb0de]::ops::function::FnOnce<()>>::call_once::{shim:vtable#0}
  46:     0x7e8412c155bb - <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once::haeea406f77e87cac
                               at /rustc/2e6fc425411280c658d852813de01c5157255088/library/alloc/src/boxed.rs:2070:9
  47:     0x7e8412c155bb - <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once::h024bb75f9bcc5dfb
                               at /rustc/2e6fc425411280c658d852813de01c5157255088/library/alloc/src/boxed.rs:2070:9
  48:     0x7e8412c155bb - std::sys::pal::unix::thread::Thread::new::thread_start::h106d07b62ccda4e0
                               at /rustc/2e6fc425411280c658d852813de01c5157255088/library/std/src/sys/pal/unix/thread.rs:108:17
  49:     0x7e84129a9ded - <unknown>
  50:     0x7e8412a2d0dc - <unknown>
  51:                0x0 - <unknown>

error: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/issues/new?labels=C-bug%2C+I-ICE%2C+T-compiler&template=ice.md

note: please make sure that you have updated to the latest nightly

note: rustc 1.81.0-nightly (2e6fc4254 2024-07-20) running on x86_64-unknown-linux-gnu

query stack during panic:
#0 [hir_crate] getting the crate HIR
#1 [hir_attrs] getting HIR owner attributes in ``
end of query stack
error: aborting due to 1 previous error

For more information about this error, try `rustc --explain E0425`.

```

</p>
</details>

### PR DESCRIPTION
**Pull request by camelid**
Fixes #128016.

Ever since #125915, some `ast::AnonConst`s turn into `hir::ConstArgKind::Path`s,
which don't have associated `DefId`s. To deal with the fact that we don't have
resolution information in `DefCollector`, we decided to implement a process
where if the anon const *appeared* to be trivial (i.e., `N` or `{ N }`), we
would avoid creating a def for it in `DefCollector`. If later, in AST lowering,
we realized it turned out to be a unit struct literal, or we were lowering it
to something that didn't use `hir::ConstArg`, we'd create its def there.

However, let's say we have a macro `m!()` that expands to a reference to a free
constant `FOO`. If we use `m!()` in the body of an anon const (e.g., `Foo<{ m!() }>`),
then in def collection, it appears to be a nontrivial anon const and we create
a def. But the macro expands to something that looks like a trivial const arg,
but is not, so in AST lowering we "fix" the mistake we assumed def collection
made and create a def for it. This causes a duplicate definition ICE.

The long-term fix for this is to delay the creation of defs for all expression-like
nodes until AST lowering (see #128844 for an incomplete attempt at this). This
would avoid issues like this one that are caused by hacky workarounds. However,
doing this uncovers a pre-existing bug with opaque types that is quite involved
to fix (see #129023).

In the meantime, this PR fixes the bug by delaying def creation for anon consts
whose bodies are macro invocations until after we expand the macro and know
what is inside it. This is accomplished by adding information to create the
anon const's def to the data in `Resolver.invocation_parents`.

r? @BoxyUwU

### CODE REVIEW

**Review by BoxyUwU — CHANGES_REQUESTED** [review]
Can you add a test for the version of the ICE that refers to an in scope generic parameter:
```rust
macro_rules! len {
    ($x:ident) => {
        $x
    };
}

fn bar<const N: usize>() {
    let val: [bool; len!(N)] = [true; N];
}
```
And also the one with `Res::Err`
```rust
macro_rules! len {
    () => {
        target
    };
}

fn main() {
    let val: [bool; len!()] = [];
}
```

**Comment by BoxyUwU** [comment]
cc @cjgillot

**Comment by petrochenkov** [comment]
Upd: I started reviewing this stuff today, starting from https://github.com/rust-lang/rust/pull/125915.
I'm generally skeptical about moving any def collection to later stages (except for leaf `DefId`s perhaps), but I need to read the relevant PRs first to say more.

**Comment by petrochenkov** [comment]
I think I disagree with both this fix and the "long term fix" from https://github.com/rust-lang/rust/issues/129247.
I have left some comments in https://github.com/rust-lang/rust/issues/129023.

https://github.com/rust-lang/rust/pull/125915 didn't need to change anything in `DefId` creation and parenting.
The `AnonConst` def ids (or whole constants) can still be created for `ConstArgKind::Path` const expressions, it doesn't mean that they should be actually used in type checking and const evaluation for those expressions.

If `min_generic_const_args` goes further and supports arbitrary paths, then there will be situations in which it's not clear whether it is a "type system constant" even at AST lowering time.
E.g. in
```rust
Type1::Assoc<Type2::Assoc>
```
we do not know whether `Type2::Assoc` is an associated type or constant (or even enum variant constructor) because it's not resolved yet, and it cannot be determined from the expected `Type1::Assoc`'s generic parameters as well because `Type1::Assoc` is not yet resolved either.

I'd prefer https://github.com/rust-lang/rust/pull/125915 to be reverted and then re-landed with the HIR representation changes (`hir::{ConstArg, ConstArgKind}`), but without the `DefId` creation and parenting changes.

**Comment by petrochenkov** [comment]
"Literature" list:
- https://hackmd.io/@BoxyUwU/BJ6_bfmD0 The plan for min_generic_const_args
- https://github.com/rust-lang/rust/pull/125915
- https://github.com/rust-lang/rust/pull/128016 
- https://github.com/rust-lang/rust/pull/128760
- https://github.com/rust-lang/rust/pull/128844
- https://github.com/rust-lang/rust/pull/128901
- https://github.com/rust-lang/rust/pull/129023
- https://github.com/rust-lang/rust/pull/129137
- https://github.com/rust-lang/rust/pull/129244
- https://github.com/rust-lang/rust/pull/129247
- https://github.com/rust-lang/rust/pull/129246

**Comment by camelid** [comment]
> #125915 didn't need to change anything in `DefId` creation and parenting.
> The `AnonConst` def ids (or whole constants) can still be created for `ConstArgKind::Path` const expressions, it doesn't mean that they should be actually used in type checking and const evaluation for those expressions.

I was under the impression that we have a HIR validation check that ensures all `DefId`s are used. Is that not right? In any case, it would seem unfortunate to create defs that don't end up being used. It's also worth pointing out that we already did (before my PR) def creation in ast_lowering for some anon consts that were not parsed as such, like const paths parsed as type paths. Finally, IIRC we need to delay def creation to properly do query feeding for anon consts, though @BoxyUwU knows much more about this than I. I'm just not sure that it's as simple as moving the def creation earlier -- though I could be wrong.

> I'd prefer #125915 to be reverted and then re-landed with the HIR representation changes (`hir::{ConstArg, ConstArgKind}`), but without the `DefId` creation and parenting changes.

If we do decide to move def creation earlier, creating unused defs in the process, then I don't see the point of reverting the whole PR. We can just make a targeted fix to the def creation parts rather than redoing the whole thing.

On that note, would you be ok with landing this _current_ PR provided that we work to figure out a longer-term solution that is effective and satisfies all parties? It would be nice to get this bug fixed on nightly without the churn of reverting #125915 and going back to square one.

**Comment by BoxyUwU** [comment]
I don't think we have such a check off the top of my head but I *do* think that creating needless DefIds is both hacky and will not work since we will have ICEs from calling queries on those DefId's which will have unexpected hir nodes associated with them (e.g. an expression instead of an anon const). This would happen when encoding queries for `AnonConst` defids.

I do not really understand the opposition to landing this PR, it gives us a fully correct DefId tree since all items defined in anon consts will have the anon const as the parent. It also does not result in having any extra `DefId`s defined which would be problematic ICE wise and also annoying when looking at debug output since we will wind up with `{constant#1}` when there is only a single constant.

I think it makes far more sense to land #129246 which will fix the immediate beta regression indefinitely, then allow this PR to be reviewed and merged which will solve things long term. As far as I can tell there is ~no downside to this PR? The only issues I've heard of with the current approach to creating defs for anon consts is that a "fully capable" def tree is desired and this PR allows for this.

I agree with @camelid that reverting the original PR seems like needless churn. IMO it is far too strong of a reaction considering the fairly minor issues with the current scheme.

**Comment by BoxyUwU** [comment]
I don't entirely know if #129247 is what we want long term. I think if we can end up with an "obviously" correct DefId tree then that's better (which this PR does).

**Comment by petrochenkov** [comment]
>I do not really understand the opposition to landing this PR

Because it layers more hacks on top of #125915's hacks and increases the amount of changes that I'd like to revert.
I'll try to implement my alternative on Saturday and see whether it can be done without introducing even worse hacks.

**Comment by cjgillot** [comment]
Reverting #125915 agrees with the general policy to revert PRs that cause high impact bugs or perf regressions, to re-land them better later. 

On the general direction, we should strive for the simpler code. This PR introduces special cases in the def collector, to accommodate special cases in ast lowering. 

This makes me wonder: is ast lowering the right place to have the special case. Or is hir->ty lowering ? 

I'd be interested in reading the design you have in mind long term for gce feature, to understand where are your tradeoffs and what constraints you battle against and should be lifted.

**Comment by BoxyUwU** [comment]
I agree that in general a policy of "revert then reland makes sense" I *tried* making a revert PR and it has merge conflicts which is why instead I filed #129246 which was easier and would also give good assurance that there shouldn't be any regressions (and also if there *is* a mistake in that PR or the revert, I think it will be easier to justify a backport of "add a feature gate check" than "touch this random compiler logic to make it not scuffed from a merge conflcit")

Moving hacks to hir ty lowering doesn't help anything IMO, I would *like* for anon consts to have their defs created there since it will help with the `type_of` feeding hack. That wouldn't change the fact that `DefCollector` shouldn't be creating the `DefId`s for things we either don't want to create or *do* want to create later on. It also is more complex since iirc `tcx.create_def` to create defids that should have bodies is a lot more error prone.

**Comment by petrochenkov** [comment]
>I don't think we have such a check off the top of my head but I *do* think that creating needless DefIds is both hacky and will not work since we will have ICEs from calling queries on those DefId's which will have unexpected hir nodes associated with them (e.g. an expression instead of an anon const). This would happen when encoding queries for `AnonConst` defids.

Always creating `DefId`s (maybe even for all generic arguments, to avoid the "we already did def creation in ast_lowering for some anon consts" bit) would be the simplest alternative.
*Not* creating `DefId`s in some cases would be an optimization that increases complexity, and as with any optimizations it should first be proven that it is beneficial.
(At metadata encoding we already know precisely which def ids actually required anon constants and which not, that knowledge can be used to skip encoding.)

>Moving hacks to hir ty lowering doesn't help anything IMO, I would *like* for anon consts to have their defs created there since it will help with the `type_of` feeding hack.

I think that would be the second viable strategy, directly opposite to the early def id creation in def collector.
We already create artificial `static`s late in compilation similarly to this (see `intern_as_new_static`).
Such nodes do not have any nontrivial HIR (all necessary information about them is fed through queries), and don't participate in any nontrivial parenting relations, neither in `HirId` nor in `DefId` sense.
They can also be created exactly when necessary, when the constant expression is complex enough, including the `Type1::Assoc<Type2::Assoc>` cases.

So in both "create def ids very early" and "create def ids very late" we won't have approximations and hacks in def collector and AST lowering.
In the first case it's because we create them for everything, and in the second case because we have full information and can create them exactly when necessary.

**Comment by camelid** [comment]
If we create defs very late, doesn't that run into the same issue of messing up parenting information for children of anon consts? I'm not sure I follow how it resolves the issue.

**Comment by petrochenkov** [comment]
If the constant is a leaf side node that doesn't parent any HIR and doesn't parent any DefIds either, like `intern_as_new_static` static, then there's no mismatch.

**Comment by BoxyUwU** [comment]
anon consts *do* parent other defs though? I don't understand the solution you're imagining. If it can be fine to create defs in hir ty lowering then I can't see how it wouldn't be possible to construct that same parent tree in ast lowering.

**Comment by petrochenkov** [comment]
I mean it's we who decide what parents what, the scheme just needs to be self-consistent and preferably convenient to work with.
Right now anonymous constant parent HIR and inner def ids, but a different scheme similar to `intern_as_new_static` is likely also possible.

As I said I see two potentially viable, but opposite schemes that are relatively simple and not hacky - https://github.com/rust-lang/rust/pull/129137#issuecomment-2307477254, with their proc and cons.
Not being integrated into def id tree is generally not good, but for anon consts specifically that maybe be not too bad, because similarly to `intern_as_new_static` statics they are fully imaginary and don't exist in source code.

**Comment by petrochenkov** [comment]
>I'll try to implement my alternative on Saturday and see whether it can be done without introducing even worse hacks.

I won't be able to do this today or tomorrow, unfortunately.
So I won't block this PR or https://github.com/rust-lang/rust/pull/129246 further, but I wrote enough details here and in https://github.com/rust-lang/rust/issues/129023 for you to think about the larger picture for `min_generic_const_args`, and maybe reconsider the direction.

**Comment by camelid** [comment]
> > I'll try to implement my alternative on Saturday and see whether it can be done without introducing even worse hacks.
> 
> I won't be able to do this today or tomorrow, unfortunately. So I won't block this PR or #129246 further, but I wrote enough details here and in #129023 for you to think about the larger picture for `min_generic_const_args`, and maybe reconsider the direction.

Okay, sounds good. I definitely agree with you on the importance of ensuring we're comfortable with the implementation and that it won't introduce more hacky code than necessary.

**Comment by BoxyUwU** [comment]
You'll need to delete the `tests/ui/feature-gates/feature-gate-const-arg-path.rs` test

**Comment by BoxyUwU** [comment]
@bors r+ rollup=never

Merging this to unblock more work on `min_generic_const_args` .

I don't consider "dont create extraneous defids" to be solely an optimisation as it will affect diagnostics output as anon const's will be misnumbered (e.g. we'll see `foo::{constant#1}` in output when there is only a single anon const). Having this increase in complexity in `DefCollector` seems logical to me since fundamentally what we are trying to do is not represent const args with anon consts in some cases that are only able to be determined *after* macro expansion.

Unconditionally creating DefIds and introducing hacks to defid encoding to skip these "fake" DefIds for anon consts might work (with the caveat that it'll make diagnostics and debugging worse(?)). However I think it places complexity in the compiler where it is not necessarily expected or relevant. 

After this lands we should be able to remove a lot of the DefId creation in ast lowering of const arguments by always representing things as `ConstArg::Path` instead of having the weird approximation in DefCollector of what needs a `DefId`. This would leave us in a final position where the only times we create `DefId`s for const args is:
- `DefCollector`
- `rustc_legacy_const_generics` handling which only knows that a function argument is actually a const argument after name resolution has been performed (unrelated but at some point we should fix the ICEs around having nested bodies in these arguments)

Hopefully once we have made it so that the DefId creation in DefCollector is not an under approximation of what defs we *actually* need the ast lowering logic will be less "hacky" and the state of all this will seem acceptable.

**Pull request merged**
